pallets/parachain-system/src/lib.rs

```diff
@@ -829,26 +847,33 @@ impl<T: Config> Pallet<T> {
 	///
 	/// If it turns out that after processing all messages the Message Queue Chain
 	/// hash doesn't match the expected.
-	fn process_inbound_downward_messages(
+	fn enqueue_inbound_downward_messages(
 		expected_dmq_mqc_head: relay_chain::Hash,
 		downward_messages: Vec<InboundDownwardMessage>,
 	) -> Weight {
 		let dm_count = downward_messages.len() as u32;
 		let mut dmq_head = <LastDmqMqcHead<T>>::get();
 
-		let mut weight_used = Weight::zero();
+		let weight_used = T::WeightInfo::enqueue_inbound_downward_messages(dm_count);
 		if dm_count != 0 {
 			Self::deposit_event(Event::DownwardMessagesReceived { count: dm_count });
-			let max_weight =
-				<ReservedDmpWeightOverride<T>>::get().unwrap_or_else(T::ReservedDmpWeight::get);
-
-			let message_iter = downward_messages
-				.into_iter()
-				.inspect(|m| {
-					dmq_head.extend_downward(m);
-				})
-				.map(|m| (m.sent_at, m.msg));
-			weight_used += T::DmpMessageHandler::handle_dmp_messages(message_iter, max_weight);
+
+			// Eagerly update the MQC head hash:
+			for m in &downward_messages {
+				dmq_head.extend_downward(m);
+			}
+			// Note: we are not using `.defensive()` here since that prints the whole value to console. In case that the message is too long, this clogs up the log quite badly.
+			let bounded =
+				downward_messages
+					.iter()
+					.filter_map(|m| match BoundedSlice::try_from(&m.msg[..]) {
+						Ok(bounded) => Some(bounded),
+						Err(_) => {
+							defensive!("Inbound Downward message was too long; dropping");
```

Comment:
We always have to call `extend_downward`, otherwise the block will get rejected. This just records that the message was "processed" by the runtime - whatever that means.  
Further there is nothing that we can do if the message is too long...